sockeye/optimizers.py

```diff
@@ -0,0 +1,202 @@
+        """
+        self.batch_state = batch_state
+
+    def pre_update_checkpoint(self, checkpoint_state: CheckpointState):
+        """
+        Called automatically at each checkpoint.
+        """
+        self.checkpoint_state = checkpoint_state
+
+    @abstractmethod
+    def update(self, index, weight, grad, state):
+        """
+        Called automatically as normal.
+        """
+        pass
+
+
+# convenience wrapper for Optimizer.Register
+register = Optimizer.register   # pylint: disable=invalid-name
+
+
+@register
+class Eve(SockeyeOptimizer):
+    """
+    The Eve optimizer is an extended version of Adam that incorporates feedback from the objective
+    function to further adapt the learning rate.
+        * "Improving Stochastic Gradient Descent with Feedback"
+          Jayanth Koushik; Hiroaki Hayashi (https://arxiv.org/abs/1611.01505)
+
+    This version allows using validation checkpoint loss in addition to training batch loss.
+
+    Eve does not currently support rescaling gradients, clipping gradients, or weight decay.
+
+    :param learning_rate: The initial learning rate.
+    :param beta1: Exponential decay rate for the first moment estimates.
+    :param beta2: Exponential decay rate for the second moment estimates.
+    :param beta3: Exponential decay rate for batch objective relative change.
+    :param beta4: Exponential decay rate for checkpoint objective relative change.
+    :param epsilon: Small value to avoid division by 0.
+    :param k_lo: Lower threshold for relative change.
+    :param k_hi: Upper threshold for relative change.
+    :param use_batch_objective: Incorporate batch objective (can use both).
+    :param use_checkpoint_objective: Incorporate checkpoint objective (can use both).
+    """
+    def __init__(self,
+                 learning_rate: float = 0.001,
+                 beta1: float = 0.9,
+                 beta2: float = 0.999,
+                 beta3: float = 0.999,
+                 beta4: float = 0.,
+                 epsilon: float = 1e-8,
+                 k_lo: float = 0.1,
+                 k_hi: float = 10,
+                 use_batch_objective: bool = True,
+                 use_checkpoint_objective: bool = False,
+                 **kwargs):
+        check_condition(any((use_batch_objective, use_checkpoint_objective)),
+                        "Must use at least one of: batch objective, checkpoint objective")
+        super().__init__(learning_rate=learning_rate, **kwargs)
+        self.beta1 = beta1
+        self.beta2 = beta2
+        self.beta3 = beta3
+        self.beta4 = beta4
+        self.epsilon = epsilon
+        self.k_lo = k_lo
+        self.k_hi = k_hi
+        self.use_batch_objective = use_batch_objective
+        self.use_checkpoint_objective = use_checkpoint_objective
+
+    def create_state(self, index: int, weight: NDArray):
+        return (zeros(weight.shape, weight.context, dtype=weight.dtype),  # mean
```

Comment:
you can use nd.zeros_like(weight, ctx=weight.context) here.
(https://mxnet.incubator.apache.org/api/python/ndarray.html#mxnet.ndarray.zeros_like)

Reply:
Nice!  This code is based on MXNet's Adam implementation which may not be totally up to date with best practice.